File: MAP/C# - Toy language/Lab10/Toy Language/Toy Language/model/expressions/ArithExp.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Toy_Language.utils;

namespace Toy_Language.model.expressions
{
    public class ArithExp : Exp
    {
        private Exp e1, e2;
        private char op;

        public ArithExp(char _op, Exp _e1, Exp _e2)
        {
            this.op = _op;
            this.e1 = _e1;
            this.e2 = _e2;
        }

        public override int Eval(IModDictionary<string, int> symbTable)
        {
            int leftExp = e1.Eval(symbTable);
            int rightExp = e2.Eval(symbTable);

            switch (op)
            {
                case '+':
                    return leftExp + rightExp;
                case '-':
                    return leftExp - rightExp;
                case '*':
                    return leftExp * rightExp;
                case '/':
                    if (rightExp == 0)
                    {
                        throw new Exception("Create new DivideZeroException class!!");
                    }
                    else
                    {
                        return leftExp / rightExp;
                    }
                default:
                    throw new Exception("Invalid op");
            }
        }

        public override string ToString()
        {
            return "" + e1 + op + e2;
        }
    }
}

File: Year 2/Semester 1/MAP/JavaRepo - Toy language/lab2/src/model/Expressions/ArithExp.java
package model.Expressions;

import exceptions.DivideByZeroException;
import exceptions.InvalidOperandException;
import exceptions.NotExistingException;
import utils.adt.MyIDictionary;
import utils.adt.MyIHeap;

public class ArithExp extends Exp {
    private Exp e1;
    private Exp e2;
    private char op;

    public ArithExp(char _op, Exp _e1, Exp _e2)
    {
        op = _op;
        e1 = _e1;
        e2 = _e2;
    }

    @Override
    public int eval(MyIDictionary<String, Integer> tbl,
                    MyIHeap<Integer, Integer> heap) throws DivideByZeroException, InvalidOperandException, NotExistingException {
        int leftExp = e1.eval(tbl, heap);
        int rightExp = e2.eval(tbl, heap);
        switch (op)
        {
            case '+': // +
                return leftExp + rightExp;
            case '-': // -
                return leftExp - rightExp;
            case '*': // *
                return leftExp * rightExp;
            case '/':
                if (rightExp == 0)
                {
                    throw new DivideByZeroException();
                }
                else
                {
                    return leftExp / rightExp;
                }
            default:
                throw new InvalidOperandException();
        }
    }

    @Override
    public String toString()
    {
        return "" + e1 + op + e2;
    }
}
